File: Day22/MinRemoveValidPar.py
def minRemoveToMakeValid(s):
    stack = []
    s = list(s)
    for i in range(len(s)):
        if s[i] == '(':
            stack.append(i)
        elif s[i] == ')':
            if stack:
                stack.pop()
            else:
                s[i] = ''
    while stack:
        s[stack.pop()] = ''
    return ''.join(s)

if __name__ == "__main__":
    s = "Lee(t(c)o)de)"
    print(minRemoveToMakeValid(s))  

File: Day22/MinRemoveValidPar.java

public class MinRemoveValidPar {
    public static String minRemoveToMakeValid(String s) {
        StringBuilder sb = new StringBuilder(s);
        java.util.Stack<Integer> stack = new java.util.Stack<>();
        for (int i = 0; i < sb.length(); i++) {
            char c = sb.charAt(i);
            if (c == '(')
                stack.push(i);
            else if (c == ')') {
                if (!stack.isEmpty())
                    stack.pop();
                else
                    sb.setCharAt(i, '*');
            }
        }
        while (!stack.isEmpty())
            sb.setCharAt(stack.pop(), '*');
        StringBuilder res = new StringBuilder();
        for (int i = 0; i < sb.length(); i++)
            if (sb.charAt(i) != '*')
                res.append(sb.charAt(i));
        return res.toString();
    }

    public static void main(String[] args) {
        String s = "lee(t(c)o)de)";
        System.out.println(minRemoveToMakeValid(s)); 
    }
}
